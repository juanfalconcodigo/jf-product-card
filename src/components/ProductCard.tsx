import React from 'react';
import styles from '../styles/styles.module.css';
import { useProduct } from '../hooks/useProduct';
import { createContext } from 'react';
import { ProductContextProps, Product, OnChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';


export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;
export interface Props {
    product: Product,
    /* children: ReactElement | ReactElement[], */
    children: (args:ProductCardHandlers) => JSX.Element,
    className?: string;
    style?: React.CSSProperties;
    onChange?: (args: OnChangeArgs) => void;
    value?: number;
    initialValues?: InitialValues
}

export function ProductCard({ product, children, className, style, onChange, value, initialValues }: Props) {
    const { count, changeCount, maxCount,reset,isCountReaced } = useProduct({ onChange, product, value, initialValues });
    return (
        <Provider value={{ count, changeCount, product, maxCount,isCountReaced }}>
            <div className={`${styles.productCard} ${className}`} style={style}>
                {children({changeCount,count,product,isCountReaced,reset,maxCount})}
            </div>
        </Provider>
    )
}
