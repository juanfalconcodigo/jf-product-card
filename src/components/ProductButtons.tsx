import React from 'react';
import { useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface Props {
    className?: string;
    style?: React.CSSProperties;
}
export const ProductButtons = ({ className, style }: Props) => {
    const { count, changeCount,isCountReaced } = useContext(ProductContext);
    

    return (
        <div className={`${styles.buttonsContainer} ${className}`} style={style}>
            <button className={styles.buttonMinus} onClick={() => changeCount(-1)}>-</button>
            <div className={styles.countLabel}>{count}</div>
            <button className={`${styles.buttonAdd} ${!isCountReaced?'':styles.disabled}`} onClick={() => changeCount(1)}>+</button>
        </div>
    )
}