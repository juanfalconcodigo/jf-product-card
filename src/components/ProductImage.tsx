import React from 'react';
import { useContext } from 'react';
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';
export interface Props {
    img?: string;
    className?: string;
    style?: React.CSSProperties;
}
export const ProductImg = ({ img, className, style }: Props) => {
    const { product } = useContext(ProductContext);
    let showImage: string;
    if (img) {
        showImage = img;
    } else if (product.img) {
        showImage = product.img;
    } else {
        showImage = noImage;
    }
    return (
        <img className={`${styles.productImg} ${className}`} src={showImage} alt="Coffe Mug" style={style} />
    )
}