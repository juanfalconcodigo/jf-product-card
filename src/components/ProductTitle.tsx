import React from 'react';
import { useContext } from 'react';
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface Props {
    className?: string;
    style?: React.CSSProperties;
    title?: string;
}
export const ProductTitle = ({ className, style, title }: Props) => {
    const { product } = useContext(ProductContext);
    return (
        <span className={`${styles.productDescription} ${className}`} style={style}>{title ? title : product.title}</span>
    )
}