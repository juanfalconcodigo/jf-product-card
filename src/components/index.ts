import { ProductButtons } from './ProductButtons';
import { ProductImg } from './ProductImage';
import { ProductTitle } from './ProductTitle';
import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductCardHOCProps } from '../interfaces/interfaces';

export { ProductButtons } from './ProductButtons';
export { ProductImg } from './ProductImage';
export { ProductTitle } from './ProductTitle';

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
    Image: ProductImg,
    Title: ProductTitle,
    Buttons: ProductButtons
});

