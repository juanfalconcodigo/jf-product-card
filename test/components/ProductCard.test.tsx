import React from "react";
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components/ProductCard';
import { product1 } from '../data/data';
const {act}=renderer;
describe('[ProductCard]', () => {
    test('devería retornar el componente', () => {
        const wrapper = renderer.create(<ProductCard product={product1}>
            {
                () => (
                    <h1>Product Card</h1>
                )
            }
        </ProductCard>);
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
    test('devería retornar del count deveria aumentar', () => {
        const wrapper = renderer.create(<ProductCard product={product1}>
            {
                ({ count, changeCount }) => (
                    <>
                        <h1>Product Card</h1>
                        <span>{count}</span>
                        <button onClick={() => changeCount(1)}></button>
                    </>
                )
            }
        </ProductCard>);
        let tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
        act(()=>{
            (tree as any).children[2].props.onClick();
        });
        tree = wrapper.toJSON();
        expect((tree as any).children[1].children[0]).toBe("1");
    })
})