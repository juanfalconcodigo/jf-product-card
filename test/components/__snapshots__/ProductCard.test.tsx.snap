// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ProductCard] devería retornar del count deveria aumentar 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`[ProductCard] devería retornar el componente 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
</div>
`;
